### ISSUE TITLE
Add type hints to improve codebase readability and enable static type checking

### ISSUE THREAD

**Issue by Niraj-Kamdar** [mindflayer/python-mocket#203]
I propose adding type hints to the mocket Python package to enhance code readability and maintainability. Type hints provide several benefits, including improved developer productivity, enhanced code documentation, and the ability to catch potential errors using static type-checking tools like Mypy.

**Comment by mindflayer** [comment]
Hi @Niraj-Kamdar, I am not sure if I got your proposal. Are you volunteering to add type hints to the project?

**Comment by mindflayer** [comment]
No answers to my question, closing it.

**Comment by mindflayer** [comment]
No answers to my question, closing it.

**Comment by Niraj-Kamdar** [comment]
Hey @mindflayer Sorry for the late reply here.  My proposal is for adding type hints in general, I would love to help here though when I get free time but currently, I have too much on my plate to help out with this. I would appreciate if you can keep this issue open so others or me can eventually take this and add type-hints to the mocket.

**Comment by Kilo59** [comment]
@mindflayer Closing this issue signals that

1. You don't plan to add type annotations.
2. You aren't willing to accept contributions that add these.

Can you re-open this issue if one or both of these are untrue?

**Comment by mindflayer** [comment]
> @mindflayer Closing this issue signals that
> 
>     1. You don't plan to add type annotations.
> 
>     2. You aren't willing to accept contributions that add these.
> 
> 
> Can you re-open this issue if one or both of these are untrue?

I am the only maintainer of Mocket, and it's worth keeping issues open only if I am confindent it's something I have time for, or at least I know someone is willing to help with.
Having a backlog of stuff just for the sake of it it's pointless.
So, allow me to bounce back the question: anyone is willing to contribute with adding type hints?

**Comment by Kilo59** [comment]
>So, allow me to bounce back the question: is anyone willing to contribute with adding type hints?

@mindflayer 

Yes. The only thing that gives me pause about taking on this work is the fact that this issue was closed.

**Comment by mindflayer** [comment]
Well, alright then! 👍

**Comment by mindflayer** [comment]
Hi @Kilo59, are you still planning to continue with this task?

**Comment by betaboon** [comment]
Hello @mindflayer,

i just ran into the same "issue" wanting to have this library typed.

I started taking a stab at it, but noticed huge parts of the core are hard to type especially due to how request- and response-types are handled (eg `Entry.response_cls`)

i think typing this library fully requires some heavy-handed refactoring in some areas.

I'd be willing to do the work (ensuring to not touch the tests), but I'm wondering if you are open to a PR that does said refactoring in addition to the typing?

kind regards
beta

**Comment by mindflayer** [comment]
Hi @betaboon, I am not against things like that, and I see many reasons for wanting type hints added to Mocket, especially for people using it as a framework for writing new mocks.
It's definitely not the main use-case, but I believe the reason is a mix of poor documentation and lack of type hints.

**Comment by betaboon** [comment]
> Hi @betaboon, I am not against things like that, and I see many reasons for wanting type hints added to Mocket, especially for people using it as a framework for writing new mocks. It's definitely not the main use-case, but I believe the reason for it is a mix of poor documentation and lack of type hints.

cool. I'll start working on a PR and push an early commit showing the direction of refactoring i have in mind, so that you can take a look and think whether you like that direction or not :)

**Comment by mindflayer** [comment]
Sure! Apart from looking like a scary monster, I am not that bad. :)

**Comment by mindflayer** [comment]
Jokes apart, it makes sense to split it in chunks, and of course I'll do my best to raise concerns as soon as I see what you have in mind.

**Comment by betaboon** [comment]
@mindflayer i just pushed a PR.

let's continue discussion there.

**Comment by JPDSousa** [comment]
Hey @mindflayer 

I have a project which would benefit from type hints in `mocket.plugins.httpretty`. I can push a PR to add them. Are you still accepting contributions here?

**Comment by mindflayer** [comment]
> I have a project which would benefit from type hints in `mocket.plugins.httpretty`. I can push a PR to add them. Are you still accepting contributions here?

Sure, feel free to send a PR.

**Comment by mindflayer** [comment]
@JPDSousa do you need a new package released with your changes?

**Comment by JPDSousa** [comment]
> @JPDSousa do you need a new package released with your changes?

That would be great, but I can wait if there's other changes planned 👍

**Comment by mindflayer** [comment]
Hi @JPDSousa, here is a new release shipping your changes:
https://pypi.org/project/mocket/3.13.6/

**Comment by mindflayer** [comment]
I finished adding type hints with the help of Copilot.
